src/schema/users.ts

```diff
@@ -1866,6 +1878,77 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
         weekStart,
       };
     },
+    recoverStreak: async (
+      _,
+      __,
+      ctx: AuthContext,
+      info,
+    ): Promise<GQLUserStreak> => {
+      const { userId } = ctx;
+      const user = await ctx.con
+        .getRepository(User)
+        .findOneByOrFail({ id: userId });
+
+      const key = generateStorageKey(
+        StorageTopic.Streak,
+        StorageKey.Reset,
+        userId,
+      );
+      const oldStreakLength = Number(await getRedisObject(key));
+      const userDoesntHaveOldStreak =
+        !oldStreakLength || !isNumber(oldStreakLength);
+      if (userDoesntHaveOldStreak) {
+        throw new ValidationError('No streak to recover');
+      }
+
+      const streak = await getUserStreakQuery(userId, ctx, info);
+      const hasNoStreakOrCurrentIsGreaterThanOne =
+        !streak || streak.current > 1;
+      if (hasNoStreakOrCurrentIsGreaterThanOne) {
+        throw new ValidationError('Time to recover streak has passed');
+      }
+
+      const lastUserRecoverAction = await ctx.con
+        .getRepository(UserStreakAction)
+        .findOneBy({
+          userId,
+          type: UserStreakActionType.Recover,
+        });
+      const isFirstRecover = !lastUserRecoverAction;
+      const recoverCost = isFirstRecover ? 0 : streakRecoverCost;
+      const userCanAfford = user.reputation >= recoverCost;
+
+      if (!userCanAfford) {
+        throw new ValidationError('Not enough reputation to recover streak');
+      }
+
+      const reputationEvent = {
+        grantToId: userId,
+        targetId: userId,
+        targetType: ReputationType.Streak,
+        reason: isFirstRecover
+          ? ReputationReason.StreakFirstRecovery
+          : ReputationReason.StreakRecover,
+        amount: recoverCost * -1,
+      };
+
+      await ctx.con.getRepository(ReputationEvent).save(reputationEvent);
+      await ctx.con.getRepository(UserStreakAction).save({
+        userId,
+        type: UserStreakActionType.Recover,
+      });
+
+      await ctx.con.getRepository(UserStreak).update(
+        {
+          userId,
+        },
+        {
+          currentStreak: oldStreakLength + streak.current,
+        },
+      );
```

Comment:
We can do this in transaction to have changes saved in atomic way in the db.

Reply:
Updated ✅